[1553] Update dependencies (May 2016)

orenyk opened the pull request:
Resolves #1553 

- Highlights:
  - Update Devise to v4.1
  - Update Capybara to v2.7
- Email validators are now more permissive due to Devise update
- Capybara matchers no longer return hidden inputs by default
- Several new rubocop cops were implemented and dealt with

orenyk commented:
Whoops forgot to update .travis.yml with the new Ruby version.

esoterik commented:
Any reason we didn't update to ruby 2.3?

orenyk commented:
Excellent point, I'll update and adjust.

orenyk commented:
Updated with Ruby 2.3.1, we can merge this once we're done with #1566.

orenyk commented:
Updated a few gems that had since pushed minor updates and rebased onto the latest `master`, resolving merge conflicts from #1538.

esoterik commented on spec/lib/csv_export_spec.rb:
```diff
@@ -19,7 +19,8 @@
       expect(csv.first.split(',')).to eq(
-        FactoryGirl.build(model).attributes.keys - PROTECTED_COLS)
+        FactoryGirl.build(model).attributes.keys - PROTECTED_COLS
+      )
```
is this preferable over this? 

```
expect(csv.first.split(',')).to \
  eq(FactoryGirl.build(model).attributes.keys - PROTECTED_COLS)
```

orenyk replied:
Yea I prefer it your way, I'll go through and update where possible.

orenyk commented:
Updated a bunch of rubocop fixes and had another update, ready for another look-through!

esoterik commented on config/initializers/00_devise.rb:
```diff
@@ -10,11 +10,11 @@
+  config.mailer_sender = if AppConfig.table_exists? && !AppConfig.first.nil?
+                           AppConfig.first.admin_email
+                         else
+                           config.mailer_sender = 'admin@reservations.app'
```
`config.mailer_sender` can be removed

orenyk replied:
Good catch!

esoterik commented on lib/equipment_import.rb:
```diff
@@ -53,8 +56,8 @@ def import_models(processed_models, model_overwrite = false)
-      if model_overwrite && (EquipmentModel.where('name = ?',
-                                                  model_data[:name]).size > 0)
+      if model_overwrite && !EquipmentModel.where('name = ?',
+                                                  model_data[:name]).empty?
```
this can be converted to `model = if ...`

esoterik commented on app/models/user.rb:
```diff
@@ -135,7 +135,7 @@ def self.search_ldap(login)
+                                    .select { |s| s && !s.empty? }.join(' ')
 
       # define username based on authentication method
       if ENV['CAS_AUTH']
```
can be converted to `out[:username] = if ...`

esoterik commented on app/models/reservation.rb:
```diff
@@ -370,56 +364,53 @@ def update(current_user, new_params, new_notes) # rubocop:disable all
+          old_val = diff[0].to_s(:long)
+          new_val = diff[1].to_s(:long)
+          if checked_out?
+            if overdue? && diff[1] >= Time.zone.today
```
can be converted to `overdue_str = if ...`

esoterik commented:
left a few small comments!

orenyk commented:
Fixed, nice catches :smile:. One more review?

esoterik commented on app/models/report.rb:
```diff
@@ -64,7 +65,7 @@ def self.item_to_row(item)
   # get the average of an array of values, discounting nil values
   def self.average2(arr)
     arr = arr.reject(&:nil?)
```
isn't this the same as `arr.compact!`?

orenyk replied:
I'm going to address this comment since it's not really included in #1570 (as compared to most of the other comments) and then we should be good to go.

esoterik commented on app/models/reservation.rb:
```diff
@@ -370,56 +364,53 @@ def update(current_user, new_params, new_notes) # rubocop:disable all
+    # write notes header
+    header = "### Edited on #{Time.zone.now.to_s(:long)} by "\
+      "#{current_user.md_link}\n"
+    self.notes = notes ? notes + "\n\n" + header : header
```
do we prefer this form of concatenation or should this be `"#{notes}\n\n#{header}"`?

esoterik commented on app/models/reservation.rb:
```diff
@@ -276,11 +272,9 @@ def to_cart
+    return 'Reservation not eligible for renewal' unless eligible_for_renew?
     self.due_date = find_renewal_date
-    self.notes = "#{notes}" + "\n\n### Renewed on "\
+    self.notes = notes.to_s + "\n\n### Renewed on "\
```
this could be `"#{notes}\n\n### Renewed on "\`, right?

esoterik commented on app/models/reservation.rb:
```diff
@@ -370,56 +364,53 @@ def update(current_user, new_params, new_notes) # rubocop:disable all
+          new_val = diff[1] ? EquipmentItem.find(diff[1]).md_link : 'nil'
         end
+        self.notes += "\n#{name} changed from " + old_val + ' to '\
+          + new_val + '.' + overdue_str.to_s
```
same as above as far as `+` vs `"#{}"`

esoterik commented on app/models/equipment_item.rb:
```diff
@@ -71,37 +71,34 @@ def make_switch_notes(old_res, new_res, handler)
+        new_val = diff[1] ? EquipmentModel.find(diff[1]).name : 'nil'
       end
+      new_notes += "\n#{name} changed from " + old_val + ' to ' + new_val\
+        + '.' if old_val && new_val
```
same as [this](https://github.com/YaleSTC/reservations/pull/1565/files#r64673683)

esoterik commented on app/models/equipment_item.rb:
```diff
@@ -71,37 +71,34 @@ def make_switch_notes(old_res, new_res, handler)
+      new_notes += "\n#{name} changed from " + old_val + ' to ' + new_val\
+        + '.' if old_val && new_val
     end
     new_notes += "\n\n" + notes
```
same as [this](https://github.com/YaleSTC/reservations/pull/1565/files#r64673683)

orenyk commented:
Updated `Report.average2` to address your comment, switched out an `if...else` with a guard clause while I was in there :smile:. If you're cool with that change then this is good to go!

esoterik commented:
yup, this is good to go

orenyk commented:
Great! I'm actually going to block this on #1566 after which I'll rebase and deal with any new rubocop errors before squashing for a final review.

orenyk commented:
Rebasing now

orenyk commented:
@esoterik ready for another glance through, there were merge conflicts in the following files:

```
spec/lib/tasks/email_*_reminder_spec.rb
spec/controllers/contact_controller_spec.rb
spec/lib/csv_export_spec.rb
```

and a new rubocop error (missing `.freeze` in `spec/controllers/reservations_controller_spec.rb`). It passed specs and the updated rubocop (so should have green builds) - let me know when you've had a chance to skim through it so we can merge.

esoterik commented:
I think this is ready!

orenyk commented:
Thanks!

orenyk commented:
Of course a few of our dependencies are already out of date, but nothing critical :stuck_out_tongue:. Let's aim to revisit this in a month (I'll open up a new issue).
